compiler/rustc_lint/src/types.rs

```diff
@@ -770,8 +773,26 @@ pub(crate) fn repr_nullable_ptr<'tcx>(
     debug!("is_repr_nullable_ptr(cx, ty = {:?})", ty);
     if let ty::Adt(ty_def, substs) = ty.kind() {
         let field_ty = match &ty_def.variants().raw[..] {
-            [var_one, var_two] => match (&var_one.fields.raw[..], &var_two.fields.raw[..]) {
+            [variant1, variant2] => match (&variant1.fields.raw[..], &variant2.fields.raw[..]) {
+                // If one variant is empty and the other is a single field (e.g. Option<T>)
                 ([], [field]) | ([field], []) => field.ty(cx.tcx, substs),
+                // If one variant has a ZST and the other has a single field (e.g. Result<T, ()> or Result<(), T>)
```

Comment:
All of these uses of "ZST" here should become "1-ZST" -- it is quite important that we exclude ZST with alignment requirements. There now is a `layout.is_1zst` method that can and should be used here.